### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by mdietrichc2c**
Odoo PR: https://github.com/odoo/odoo/pull/12645

Description of the issue/feature this PR addresses:
Basic performance improvements for 2 issues, when migration a client with ~2 millions analytic lines and ~4 million sale order lines:
- displaying analytic accounts with computed debit/credit/balance,
- validating an invoice with analytic information.

Current behavior before PR:
- after 15 minutes, analytic accounts are still not displayed.
- validating an invoice takes 10-12 minutes.

Desired behavior after PR is merged:
- displaying analytic accounts takes 1 second.
- validating an invoice still takes 10-15 seconds, but it's still some improvement.
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by jgrandguillaume** [comment]
Thanks for the contribs ! Code review :+1:

File: addons/analytic/models/analytic.py [hunk 1]
```diff
@@ -19,27 +19,25 @@ class account_analytic_account(models.Model):
...
-        domain = [('account_id', 'in', self.mapped('id'))]
+        sql_query = ("""
+            SELECT account_id, SUM(amount)
+            FROM account_analytic_line
+            WHERE account_id IN %s
+            AND company_id = %s
+        """ % (tuple(self.ids), self.env.user.company_id.id))
         if self._context.get('from_date', False):
-            domain.append(('date', '>=', self._context['from_date']))
+            sql_query += (""" AND date >= %s""" % self._context['from_date'])
```
Comment (Garamotte) [inline addons/analytic/models/analytic.py]: Please use cr.execute arguments, to avoid SQL injections

File: addons/analytic/models/analytic.py [hunk 2]
```diff
@@ -19,27 +19,25 @@ class account_analytic_account(models.Model):
...
-        account_amounts = analytic_line_obj.search_read(domain, ['account_id', 'amount'])
-        account_ids = set([line['account_id'][0] for line in account_amounts])
-        data_debit = {account_id: 0.0 for account_id in account_ids}
-        data_credit = {account_id: 0.0 for account_id in account_ids}
-        for account_amount in account_amounts:
-            if account_amount['amount'] < 0.0:
-                data_debit[account_amount['account_id'][0]] += account_amount['amount']
-            else:
-                data_credit[account_amount['account_id'][0]] += account_amount['amount']
+            sql_query += (""" AND date <= %s""" % self._context['to_date'])
```
Comment (Garamotte) [inline addons/analytic/models/analytic.py]: Please use cr.execute arguments, to avoid SQL injections

**Comment by mdietrichc2c** [comment]
@sylvain-garancher thanks, I missed that part. It's now fixed.

**Comment by StefanRijnhart** [comment]
thanks. Just a bit wary that _everything_ that involves millions of records can be improved by replacing ORM implementations with SQL, but we'll have to see!

-- edit: retract thumbs up as per comment below --

**Comment by StefanRijnhart** [comment]
@sylvain-garancher is that a thumbsup now?

**Comment by Garamotte** [comment]
Sorry, didn't test, I just read the code.

**Comment by gdgellatly** [comment]
Hi,

I'm not necessarily convinced by this.  I have not tested, but have done a lot of similar performance improvements.  Firstly the index and analytic storage on Sale Order line.  All good by me.  Its the SQL query that kind of interests me as I try to avoid them wherever possible.

Firstly I thought %s had to be escaped like %%s - maybe I'm wrong, been a while but thought I'd mention.

The ORM search it replaces does nothing more than really add a group by to the existing search.  

However having done a lot of these, the issue causing the most performance problems would be the for loop and the += statements.  This is a really dumb way to sum items in a dictionary.  What follows is not criticism of code, but rather a different angle on the issue.

In the standard library we have itertools.groupby() and I am a big fan of the toolz library as well for similar operations.  I'd be really interested to know if just replacing the for loop with itertools.groupby() would alleviate most of the issue.

But of course the best solution to this is not pure SQL at all, because it is such a common issue (grouping ORM search results).  In master it would be great to replace the signature of the dependent methods (search, search_read, _search) and implementation to support groupby sql operations.  Just seems this fixes a sympton, one I've seen too often, when the cause is lack of ORM support for group by.

EDITS: I made some edits as I misread part.
EDITS: again I see it is already in master support for this - def read_group - surely this must be correct solution

**Comment by gdgellatly** [comment]
guess I should have pinged people.  Had another day to look at and not writing at 2am in the morning.   While I have no doubt the code achieves what it is meant to, definitely I think read_group is a far more suitable solution.

@mdietrichc2c @StefanRijnhart @jgrandguillaume

**Comment by StefanRijnhart** [comment]
Obviously, if performance can be improved by ORM methods such as read_group that would be preferred. @mdietrichc2c, the code should not be too hard. A benchmark of both versions would greatly help.

**Comment by mdietrichc2c** [comment]
@jgrandguillaume @StefanRijnhart @gdgellatly : sorry for the delay, I did not see the responses until today.
I modified the code to use `read_group()` instead of the SQL request. On the same conditions as the original improvement, it took 2-3 seconds to display the whole analytic account list.

**Comment by gdgellatly** [comment]
@mdietrichc2c awesome thanks.  I know its not as quick as original but it is my preferred.  I think most of the doubling in time is primarily due to the second read that occurs for the m2o name operation and then the iteration to strip the read name. 

But 2-3 seconds is better than never displaying, and we can always work on getting an option into master for not calling the many2one second read.  Is kind of an annoying behavior anyway.  Usually you need one or the other (id or name), rarely both.

:+1:  for me - code review - no test

**Comment by mdietrichc2c** [comment]
@gdgellatly no, I'm okay with the trade-off as well ; it's always better to use ORM methods than SQL.

I see that the build is red, but it seems unrelated.

File: addons/analytic/models/analytic.py [hunk 3]
```diff
@@ -19,27 +19,22 @@ class account_analytic_account(models.Model):
...
+                       for amount in amounts])
 
         for account in self:
-            account.debit = abs(data_debit.get(account.id, 0.0))
-            account.credit = data_credit.get(account.id, 0.0)
-            account.balance = account.credit - account.debit
+            amount = amounts.get(account.id, 0.0)
+            account.balance = amount
+            account.credit = amount if amount > 0.0 else 0.0
+            account.debit = -amount if amount < 0.0 else 0.0
```
Comment (StefanRijnhart) [inline addons/analytic/models/analytic.py]: Oh wait, I think I missed this. It seems you are replacing the sums of debit and credit amounts with the balance. That is not good. What you typically like to see is the debit and credit of an account which represent the expenses and income, and then the balance which represents the profit and loss. By summing debit and credit amongst each other we lose the first type of information.

**Comment by mdietrichc2c** [comment]
@StefanRijnhart fixed it (still a new hit on performances, but since it's needed, no other option)

**Comment by gdgellatly** [comment]
This now confuses me. Why would debits be negative and credits positive.  I
seem to recall about 5 years ago Odoo had the terms confused in some
places, maybe this was one. Will look closer when I get to work.

**Comment by gdgellatly** [comment]
@StefanRijnhart @mdietrichc2c I've reviewed the upstream code and tried to piece together.  I don't really use analytics but from what I can tell this whole piece of code should be incorrect.  

Simply put the balance is correct because the equation is credit-debit.  However the correct formula for a balance is debit-credit (I'm shortcutting a little here to avoid indepth discussion on debit and credit balances vs how Odoo records figures which is cr balances as negative).  

Therefore the amounts displaying in the debits column is actually the credits and vice versa.  Its not a bug with your code, but while its being fixed, then I don't know whether this should be fixed as well.

Edit for confirmation of the above assertion, lines from account_move.py
https://github.com/OCA/OCB/blob/9.0/addons/account/models/account_move.py#L248
https://github.com/OCA/OCB/blob/9.0/addons/account/models/account_move.py#L291 - probably best example as doesn't use abs.
https://github.com/OCA/OCB/blob/9.0/addons/account/models/account_move.py#L350
https://github.com/OCA/OCB/blob/9.0/addons/account/models/account_move.py#L694

**Comment by mdietrichc2c** [comment]
@gdgellatly : I think it's ok to keep the same behavior as before in this PR ; if Odoo changes debit/credit later on, it will be easier to change it in another PR at a later date.

**Comment by StefanRijnhart** [comment]
:+1: thanks for the follow up!

File: addons/analytic/models/analytic.py [hunk 4]
```diff
@@ -19,26 +19,28 @@ class account_analytic_account(models.Model):
...
-            if account_amount['amount'] < 0.0:
-                data_debit[account_amount['account_id'][0]] += account_amount['amount']
-            else:
-                data_credit[account_amount['account_id'][0]] += account_amount['amount']
+            domain += [('date', '<=', self._context['to_date'])]
+        # compute debits
+        debit_domain = domain + [('amount', '<', 0.0)]
+        debits = self.env['account.analytic.line'].read_group(
+            debit_domain, ['account_id', 'amount'], ['account_id'])
+        debits = dict([(amount['account_id'][0], amount['amount'])
```
Comment (mistotebe) [inline addons/analytic/models/analytic.py]: If you use a dictionary comprehension like `debits = {amount['account_id'][0]: amount['amount'] for amount in debits}` you might save some more time (admittedly, not much, but it also helps with readability):

```
$ ipython
In [1]: import random

In [2]: l = range(1000)

In [3]: random.shuffle(l)

In [4]: %timeit d = dict((x, x+3) for x in l)
1000 loops, best of 3: 250 µs per loop

In [5]: %timeit d = {x: x+3 for x in l}
10000 loops, best of 3: 151 µs per loop

In [6]: l = range(100000)

In [7]: random.shuffle(l)

In [8]: %timeit d = dict((x, x+3) for x in l)
10 loops, best of 3: 76.5 ms per loop

In [9]: %timeit d = {x: x+3 for x in l}
10 loops, best of 3: 64.3 ms per loop
```

File: addons/analytic/models/analytic.py [hunk 5]
```diff
@@ -19,26 +19,28 @@ class account_analytic_account(models.Model):
...
+                       for amount in debits])
+        # compute credits
+        credit_domain = domain + [('amount', '>', 0.0)]
+        credits = self.env['account.analytic.line'].read_group(
+            credit_domain, ['account_id', 'amount'], ['account_id'])
+        credits = dict([(amount['account_id'][0], amount['amount'])
```
Comment (mistotebe) [inline addons/analytic/models/analytic.py]: Same here

**Comment by mdietrichc2c** [comment]
@mistotebe true, I'm used to develop in 7.0 (where the minimum Python requirement was 2.6, hence no dict comprehension).
It's now changed.

**Comment by mistotebe** [comment]
LGTM

**Comment by damdam-s** [comment]
:+1: thanks

**Comment by yvaucher** [comment]
:+1: Thanks for this improvement @mdietrichc2c we can now merge it.

**Pull request merged**